### PR TITLE
o/snapstate: ignore remove errors in clear-snap handler, only log them

### PR DESCRIPTION
**Pull request by stolowski**
Do not fail snap remove due to errors when cleaning up snap data - only log the errors on the tasks.

This avoids issues where a snap may lock (and not release) a resource inside its snap data folder which prevents uninstalling; the only known case of this so far was lxd snap with a bug in remove hook (already fixed in lxd). It seems sensible to allow snap remove to complete the operation in such case and just warn the user about leftover data - after all, depending on when during the os.Remove did it fail, some or all data could have been removed already

### CODE REVIEW

File: overlord/snapstate/handlers.go [hunk 1]
```diff
@@ -1894,13 +1894,17 @@ func (m *SnapManager) doClearSnapData(t *state.Task, _ *tomb.Tomb) error {
 	}
 
 	if err = m.backend.RemoveSnapData(info); err != nil {
-		return err
+		st.Lock()
+		t.Logf("Cannot remove snap data: %v", err)
```
Comment (zyga) [inline overlord/snapstate/handlers.go:1898]: This seems like a perfect place to warn instead.

**Review by zyga — COMMENTED** [review]
Quick opinion.

**Review by mvo5 — APPROVED** [review]
Thank you

File: overlord/snapstate/handlers.go (hunk 1 above)
Comment (pedronis) [reply overlord/snapstate/handlers.go:1898]: I thought a bit and I think I agree,  though in practice I thing we should do both

File: overlord/snapstate/snapstate_remove_test.go [hunk 2]
```diff
@@ -1426,12 +1427,18 @@ name: foo
...
+		st.OkayWarnings(time.Now())
+
 		st.Unlock()
 		s.settle(c)
 		st.Lock()
 
 		c.Check(chg.Status(), Equals, state.DoneStatus)
 		c.Check(strings.Join(t.Log(), ""), testutil.Contains, tc.expectedMessage)
+
+		warnings, _ := st.PendingWarnings()
```
Comment (mvo5) [inline overlord/snapstate/snapstate_remove_test.go:1439]: Nice!

**Review by pedronis — COMMENTED** [review]
The code looks fine, but 

1) I don't think we want this in 2.48, earliest 2.49
2) It's a bit of a backward incompatible behavior change
3) with this change "snap remove" will exit with status 0 but the snap has still bits around

We might need to do a bit more around this change, I will mark it blocked for now.

**Comment by anonymouse64** [comment]
For what it's worth, I encountered another snap that runs into this issue, microk8s, a classic snap, creates systemd mounts under /var/snap/microk8s/common/var/... and it seems to suffer from a bug where these mount units aren't unmounted properly sometimes, thus failing snap remove, see https://github.com/ubuntu/microk8s/issues/1941. There I think the real root cause of the bug is microk8s not cleaning up after itself, but just thought I would mention it as it is similar to the LXD case

**Comment by anonymouse64** [comment]
What's the status of this?

**Comment by stolowski** [comment]
> What's the status of this?

Per @pedronis comments there are aspects to reconsider. We probably should discuss what would be the expected behavior first.

**Comment by pedronis** [comment]
we might come back to this one at a later point

**Pull request closed without merging**